### ISSUE TITLE
Allow using Content-Security-Policy without unsafe-inline

### ISSUE THREAD

**Issue by chrisdavidmiles** [beaverbuilder/assistant#140]
This is not really a bug, but more of a feature request. This plugin makes use of inline JavaScript that would be blocked by any Content Security Policy not using `unsafe-inline` directive. It's not possible to use a secure CSP and use this plugin in it's current state. And typical CSP workarounds don't seem practical in the case since:
- Creating an exception hash wouldn't work more than once as `FL_ASSISTANT_CONFIG` is highly dynamic.
- Creating an exception via token could work, but would likely require a level of "catch-all" token generation that would defeat the purpose of using a CSP in the first place.
- Falling back to the `unsafe-inline` directive would also defeat the purpose of using a CSP.

If this plugin didn't use inline JavaScript, and instead made use of an external js resource, perhaps through admin-ajax, CSP wouldn't be an issue at all. While this does introduce some additional complexity to the plugin, I believe the security benefit would be worth it.

**Comment by brentjett** [comment]
That's an interesting case. Does WordPress itself not also use some inline scripts (WP.emoji i think?) ? 
I'm fine with making a request for most of what's inside our config object, but there are a few pieces of info we need to know to be able to get the panel rendered pretty quickly after page load, like whether the panel is currently showing or not. Any suggestions on other ways to ship some basic data down with the page?

**Comment by chrisdavidmiles** [comment]
Yes, WordPress core does print the emoji js to the page, but all of that can be disabled or made not to be inline without breaking since it doesn't change. 

And what about stuff you want to happen inline like "like whether the panel is currently showing or not"? As long as whatever is inline is predictable and won't often change, and there's not too many of them, then it could feasibly be hashed as an exception to CSP. The only place this wouldn't make sense would be the config object since that's going to be different every time and not feasible to hash ahead of time. But if it's something that's easy to hash, and it won't often change in future releases, then it's reasonable to use that. If you wanted to go the extra mile you could even publish the hashes to a docs page as plugin authors often do. "How to make [plugin name] work with CSP" so that site owners don't have to figure out the hashes every time themselves.

And if you use admin-ajax for the config parts, you would be able to get around the caching issues of making dynamic js an external resource, since any hosting stack running WordPress would already need to be whitelisting that.

As far as how many folks are actually going to care about CSP, I'd say it's higher than you might expect, but it's also a group that's used to going through the hassle of making it work with WordPress. Many popular plugins like Jetpack still require some hash exceptions to work properly. And this will likely continue, since `wp_add_inline_script` is so easy to use. 

But if at least the highly dynamic stuff in a plugin isn't printed insline, then the problem goes from impossible for site owners, to doable for site owners.

**Comment by chrisdavidmiles** [comment]
Yes, WordPress core does print the emoji js to the page, but all of that can be disabled or made not to be inline without breaking since it doesn't change.

And what about stuff you want to happen inline like "like whether the panel is currently showing or not"? As long as whatever is inline is predictable and won't often change, and there's not too many of them, then it could feasibly be hashed as an exception to CSP. The only place this wouldn't make sense would be the config object since that's going to be different every time and not feasible to hash ahead of time. But if it's something that's easy to hash, and it won't often change in future releases, then it's reasonable to use that. If you wanted to go the extra mile you could even publish the hashes to a docs page as plugin authors often do. "How to make [plugin name] work with CSP" so that site owners don't have to figure out the hashes every time themselves.

And if you use admin-ajax for the config parts, you would be able to get around the caching issues of making dynamic js an external resource, since any hosting stack running WordPress would already need to be whitelisting that.

As far as how many folks are actually going to care about CSP, I'd say it's higher than you might expect, but it's also a group that's used to going through the hassle of making it work with WordPress. Many popular plugins like Jetpack still require some hash exceptions to work properly. And this will likely continue, since `wp_add_inline_script` is so easy to use.

But if at least the highly dynamic stuff in a plugin isn't printed inline, then the problem goes from impossible for site owners, to doable for site owners.

**Comment by markman77** [comment]
@brentjett @fastlinemedia @chrisdavidmiles Has there been any progress with this thread? I have a client requesting a CSP, but I am not able to get much of my styling to load correctly and I am struggling to get this to work.
